config/decoder.go

```diff
@@ -205,9 +209,18 @@ func (d *decoder) mapping(childKey string, value reflect.Value, def string) erro
 	// child yamlNode parsed from yaml file is of type map[interface{}]interface{}
 	// type casting here makes sure that we are iterating over a parsed map.
 	if v, ok := val.(map[interface{}]interface{}); ok {
-		childKey += _separator
+		if childKey != "" {
+			childKey += _separator
+		}
+
 		for key := range v {
 			mapKey := childKey + fmt.Sprintf("%v", key)
+
+			// Skip empty keys, because lookups in provider can toggle infinite loops/spill internals.
+			if mapKey == "" {
```

Comment:
Actually you right, it is an error, because it leads to ambiguity between root element and element with an empty key.